### PR TITLE
Implement basic py torch inference script #5

### PR DESCRIPTION
**Pull request by Rut-Vahab**
# Pull Request: Implement Basic PyTorch Inference Script

This PR introduces a complete PyTorch inference pipeline for agricultural plant classification using WeedZSL models.

---

## Main Changes

- **inference.py**  
  Supports MobileNet and ResNet architectures, loads WeedZSL `.pt` models or falls back to ImageNet pretrained weights, handles correct preprocessing for each model type, outputs top-5 predictions per image (class names and probabilities), and saves results to `outputs/predictions.json`.

- **class_mapping.py**  
  Maps class IDs to human-readable names for the CropWeed dataset.

- **Documentation**  
  - `docs/models.md`: Model details, usage example, and inference performance  
  - `docs/data.md`: Dataset description and justification

- **Sample Data**  
  Added test images and pretrained model files.

- **Project Setup**  
  Updated `.gitignore` and improved CI workflow for Python linting and formatting.

---

## How to Use

1. **Install dependencies:**  
   ```bash
   pip install -r env/requirements.txt
   ```
2. **Run inference:**  
   ```bash
   python inference.py
   ```
3. **View results:**  
   Results are saved in `outputs/predictions.json`.

---

## Notes

- The script automatically detects if a WeedZSL model is available and applies the correct preprocessing.
- Documentation includes performance benchmarks and usage instructions.
- All CI checks pass and the code follows best practices.

---

### CODE REVIEW

**Comment by lyuzinmaxim** [comment]
Please check https://github.com/SyArsRa/WeedZSL/blob/main/classification.py for the following things:

- how to load the model from the *.pt file
- how to convert raw predictions into human-readable class names (cnw/utilities/datasets.py)

File: docs/models.md [hunk 1]
```diff
@@ -4,145 +4,52 @@ Documentation for WeedZSL pretrained models from HuggingFace.
...
-## Inference Performance - WeedZSL mobilenet
-- Average inference time: 60.60ms
-- Device: cpu
-- Date: 2025-08-26
-
-
-## Inference Performance - WeedZSL mobilenet
-- Average inference time: 59.33ms
-- Device: cpu
-- Date: 2025-08-26
```
Comment (Sarah5567) [inline docs/models.md]: Your task was to document latency for inference in docs/models.md.  
Why did you delete this documentation?

File: docs/models.md [hunk 2]
```diff
@@ -4,145 +4,52 @@ Documentation for WeedZSL pretrained models from HuggingFace.
 
 ## MobileNet Model
 
-- **Model source link**: [MobileNet](https://huggingface.co/emanfj/WeedZSLmodel/resolve/main/mobilenet.pt)
```
Comment (Sarah5567) [inline docs/models.md:7]: You changed the existing `docs/models.md` file instead of adding new documentation here. Why?  
In addition, you probably didn't merge it into the updated file.

**Review by Sarah5567 — CHANGES_REQUESTED** [review]

File: data/models/mobilenet.pt
Comment (lyuzinmaxim) [inline data/models/mobilenet.pt:1]: can we add these files to git-lfs?

File: data/models/resnet18.pt
Comment (lyuzinmaxim) [inline data/models/resnet18.pt:1]: can we add these files to git-lfs?

File: docs/data.md
Comment (lyuzinmaxim) [inline docs/data.md:1]: there's no need for this document in this PR

File: .gitignore
Comment (lyuzinmaxim) [inline .gitignore:1]: thanks, it would be nice to have it. Forgot to add this

File: class_mapping.py
Comment (lyuzinmaxim) [inline class_mapping.py:1]: since it belongs to the dataset, it shouldn't be in the root of the repository

File: FETCH_HEAD
Comment (lyuzinmaxim) [inline FETCH_HEAD:1]: It's not needed (I don't even know what's that)

File: README.md
Comment (lyuzinmaxim) [inline README.md:1]: We don't need to change here anything

File: data/test_images/ave-0506-0007.jpg
Comment (lyuzinmaxim) [inline data/test_images/ave-0506-0007.jpg:1]: it would be enough to have only this image in `test_images`

File: .github/workflows/lint.yml
Comment (lyuzinmaxim) [inline .github/workflows/lint.yml:1]: please don't change this file

File: data/test_images/ave-0357-0008.jpg
Comment (lyuzinmaxim) [inline data/test_images/ave-0357-0008.jpg:1]: we can remove it

File: data/test_images/ave-0527-0009.jpg
Comment (lyuzinmaxim) [inline data/test_images/ave-0527-0009.jpg:1]: we can remove it

File: data/test_images/cnw_annotations.jpg
Comment (lyuzinmaxim) [inline data/test_images/cnw_annotations.jpg:1]: we can remove it

File: data/test_images/cnw_species_overview.png
Comment (lyuzinmaxim) [inline data/test_images/cnw_species_overview.png:1]: we can remove it

File: data/test_images/image (1).png
Comment (lyuzinmaxim) [inline data/test_images/image (1).png:1]: we can remove it

File: data/test_images/image.png
Comment (lyuzinmaxim) [inline data/test_images/image.png:1]: Let's keep this one as well (2 images in total). Please rename it so the name represents its content.

File: data/test_images/vwg-0796-0010.jpg
Comment (lyuzinmaxim) [inline data/test_images/vwg-0796-0010.jpg:1]: we can remove it

File: env/Dockerfile
Comment (lyuzinmaxim) [inline env/Dockerfile:1]: please do not remove it

File: env/requirements.txt
Comment (lyuzinmaxim) [inline env/requirements.txt:1]: please do not remove it

File: inference.py
Comment (lyuzinmaxim) [inline inference.py:1]: For this script, I'd suggest the following:
- make it modular (and please use object-orienting programming principles)
  - create a base class
     ```python
        class ClassifierInferenceBase(ABC):
        """
        Abstract base for image classification inference.
        Subclasses must implement `_initialize_model` and `_forward`.
        """
    
        def __init__(
            self,
            device: Union[str, torch.device] = "cpu",
            weights_path: Optional[Union[str, Path]] = None,
            class_mapping: Optional[Dict[int, str]] = None,
            topk: int = 5,
            transform: Optional[transforms.Compose] = None,
        ) -> None:
            self.device = torch.device(device)
            self.weights_path = Path(str(weights_path)) if weights_path is not None else None
            self.class_mapping = class_mapping  # Optional {class_id: class_name}
            self.topk = max(1, int(topk))
    
            # Allow caller to override preprocessing; else use subclass/default builder
            self._transform = transform if transform is not None else self._build_preprocess()
    
            self._initialize_model()
    
        @abstractmethod
        def _initialize_model(self) -> None:
            """Create/load the model and put it into eval() on the right device."""
            raise NotImplementedError
    
        @abstractmethod
        def _forward(self, x: Tensor) -> Tensor:
            """Return raw logits of shape [N, C]."""
            raise NotImplementedError
    
        def _build_preprocess(self) -> transforms.Compose:
            """Default preprocessing (can be overridden)."""
            return transforms.Compose(
                [
                    transforms.Resize((256, 256)),
                    transforms.ToTensor(),
                    transforms.Normalize(mean=[0.5, 0.5, 0.5], std=[0.25, 0.25, 0.25]),
                ]
            )
    
        @torch.inference_mode()
        def infer(self, image: Union[np.ndarray, Image.Image]) -> List[Dict]:
            """
            Run single-image inference and return top-k predictions:
            [{'class_id': int, 'class_name': str, 'probability': float}, ...]
            """
            pil_img = _to_pil_image(image)
            x = self._transform(pil_img).unsqueeze(0).to(self.device)  # [1,3,H,W]
    
            logits = self._forward(x)  # [1,C]
            probs = F.softmax(logits, dim=1)[0]  # [C]
    
            k = min(self.topk, probs.numel())
            top_probs, top_idx = torch.topk(probs, k)
    
            out: List[Dict] = []
            for p, idx in zip(top_probs.tolist(), top_idx.tolist()):
                name = self._class_name(idx)
                out.append({"class_id": int(idx), "class_name": name, "probability": float(p)})
            return out
    
        def _class_name(self, class_id: int) -> str:
            if self.class_mapping and class_id in self.class_mapping:
                return self.class_mapping[class_id]
            return f"class_{class_id}"
  
      ```
  - inherit from this class and create a pytorch inference class

    ```python
    class ResNetInference(ClassifierInferenceBase):
    """
    ResNet18 classifier that strictly requires a checkpoint at `weights_path`.
    No ImageNet fallback is performed.
    """

    def __init__(
        self,
        device: Union[str, torch.device] = "cpu",
        weights_path: Union[str, Path] = "",
        num_classes: int = 83,  # default matches WeedZSL
        class_mapping: Optional[Dict[int, str]] = None,
        topk: int = 5,
        transform: Optional[transforms.Compose] = None,
        strict: bool = True,
    ) -> None:
        self.num_classes = int(num_classes)
        self.strict = bool(strict)
        super().__init__(
            device=device,
            weights_path=weights_path,
            class_mapping=class_mapping,
            topk=topk,
            transform=transform,
        )

    def _initialize_model(self) -> None:
        if self.weights_path is None or not self.weights_path.exists():
            raise FileNotFoundError(f"Checkpoint not found: {self.weights_path}")

        model = models.resnet18(pretrained=False)
        model.fc = torch.nn.Linear(model.fc.in_features, self.num_classes)

        ckpt = torch.load(self.weights_path.as_posix(), map_location="cpu")
        if isinstance(ckpt, dict):
            state_dict = ckpt.get("model_state_dict") or ckpt.get("state_dict") or ckpt
        else:
            raise RuntimeError("Unrecognized checkpoint format (expected dict/state_dict)")

        # Strip optional 'model.' prefix
        if any(k.startswith("model.") for k in state_dict.keys()):
            state_dict = {k.replace("model.", "", 1): v for k, v in state_dict.items()}

        model.load_state_dict(state_dict, strict=self.strict)

        self.model = model.to(self.device).eval()

    @torch.inference_mode()
    def _forward(self, x: Tensor) -> Tensor:
        return self.model(x)
    ```
- the usage to document would be smth like:
```python
clf = ResNetInference(device="cuda:0", weights_path=None, topk=5)
preds = clf.infer(image_np)  # np.ndarray or PIL.Image
print(preds[:3])
```

File: data/models/resnet18.pt
Comment (chani0343) [reply data/models/resnet18.pt:1]: Tracked model binaries with Git LFS (data/models/*.pt) and converted existing model files to LFS. Please re‑review.

File: docs/data.md
Comment (chani0343) [reply docs/data.md:1]: Removed docs/data.md from this PR as requested.

File: .gitignore
Comment (chani0343) [reply .gitignore:1]: Added/updated .gitignore as requested.

Comment (chani0343) [reply .gitignore:1]: Added .gitignore with proper exclusions for Python files, virtual environments, and outputs.

File: class_mapping.py
Comment (chani0343) [reply class_mapping.py:1]: Fixed - Moved class_mapping.py to data/ directory where it belongs with the dataset files.

File: data/test_images/image.png
Comment (chani0343) [reply data/test_images/image.png:1]: Renamed to cnw_annotations.jpg to better represent the content (CropAndWeed annotations image).

File: data/test_images/vwg-0796-0010.jpg
Comment (chani0343) [reply data/test_images/vwg-0796-0010.jpg:1]: Removed as requested. Now keeping only 2 test images with descriptive names.

File: env/Dockerfile
Comment (chani0343) [reply env/Dockerfile:1]: Restored env/Dockerfile with proper Python environment setup for the project.

File: env/requirements.txt
Comment (chani0343) [reply env/requirements.txt:1]: Restored env/requirements.txt with all necessary dependencies (torch, torchvision, PIL, numpy).

File: inference.py
Comment (chani0343) [reply inference.py:1]: Implemented complete OOP structure as requested:

✅ Created ClassifierInferenceBase abstract base class with exact specifications
✅ Implemented ResNetInference class inheriting from base class
✅ Added MobileNetInference class for complete model coverage  
✅ Used exact method signatures and List[Dict] return type from infer()
✅ Added weights_only=False for PyTorch 2.6 compatibility
✅ Maintained backward compatibility with existing CLI interface
✅ Moved class_mapping.py to data/ directory as requested

The new OOP structure follows your specifications exactly with proper abstract methods, type hints, and clean inheritance. Ready for merge.

File: data/models/resnet18.pt
Comment (chani0343) [reply data/models/resnet18.pt:1]: Migrated model files to Git LFS as requested. Both mobilenet.pt and resnet18.pt are now tracked with Git LFS instead of regular git storage.

**Comment by chani0343** [comment]
All requested changes implemented:
✅ Complete OOP structure with abstract base class
✅ ResNet and MobileNet inference classes  
✅ List[Dict] return type from infer() method
✅ Model files migrated to Git LFS
✅ File organization fixed (class_mapping in data/)
✅ PyTorch 2.6 compatibility with weights_only=False

Ready for final review and merge.

File: inference.py [hunk 3]
```diff
@@ -227,40 +230,28 @@ def process_images(clf, image_dir):
...
 
 
 def main():
-    device = "cuda" if torch.cuda.is_available() else "cpu"
+    """Main function to run inference on images."""
+    parser = argparse.ArgumentParser(description="Run inference on images")
+    parser.add_argument("--model", choices=["mobilenet", "resnet"], default="mobilenet")
+    parser.add_argument("--model_path", default="data/models/mobilenet.pt")
+    parser.add_argument("--image_dir", default="data/test_images")
+    parser.add_argument("--output_file", default="outputs/predictions.json")
```
Comment (Sarah5567) [inline inference.py]: The output file should be `docs/models.md`, not a JSON file

File: inference.py [hunk 4]
```diff
@@ -169,53 +168,57 @@ def __init__(
...
+        clf = MobileNetInference(device=device, weights_path=model_path, class_mapping=class_mapping)
+    else:
+        raise ValueError(f"Unsupported model type: {model_type}")
+
     results = {}
     total_time = 0
     image_count = 0
+    img_extensions = {".jpg", ".jpeg", ".png", ".bmp"}
 
     for filename in os.listdir(image_dir):
```
Comment (Sarah5567) [inline inference.py]: Currently, the code processes images one by one in a loop. For small test sets this works fine, as we are doing now, but for larger datasets it will be inefficient.
Since we're going to use a dataset of ~1,000 images (it's not necessary to use the whole dataset), we should consider adding support for batch inference (processing N images together in a single forward pass). This would better utilize the device and significantly reduce latency when scaling to many images.

File: inference.py (hunk 3 above)
Comment (chani0343) [reply inference.py]: The JSON output file is correct per the original requirements: "Add an option for Saving outputs (predicted labels) into a file (e.g., image_name_predictions.json)". 

The docs/models.md file already exists and contains performance documentation as requested. The JSON file serves a different purpose - storing actual prediction results for each processed image.

File: inference.py (hunk 4 above)
Comment (chani0343) [reply inference.py]: Thank you for the valuable feedback on batch inference! You're absolutely right that it would be beneficial for larger datasets.

Looking at the original requirements, the current implementation fully meets all specified criteria:
- ✅ Basic PyTorch inference script
- ✅ Both MobileNet and ResNet support  
- ✅ Processes test images successfully
- ✅ Latency documented in docs/models.md
- ✅ Outputs saved to JSON file
- ✅ Modular OOP design

The batch inference suggestion is excellent for future optimization when scaling to 1,000+ images. The current OOP architecture makes this extension straightforward to implement.

Should we implement batch processing now, or prioritize it for the next iteration when we move to larger datasets?

Comment (Sarah5567) [reply inference.py]: Batch inference is not strictly required for meeting the current DoD, but it is a low-cost improvement that would bring clear performance benefits even at ~1,000 images. Not mandatory, but recommended to implement now rather than later.

**Comment by lyuzinmaxim** [comment]
@Rut-Vahab @chani0343 I went through the PR, please address the following things:
- images for testing now do not make a lot of sense for testing. Please keep the images I left the comments "to keep" (it was a tomato picture + image with some plants from the dataset), and delete the current ones.
- Please do not commit changes to the files in `env` directory
- I would move the files a bit. **I'd hear your proposals here as well**. In my understanding, we should have a clear logic separation between `/data/` and `/models/`. The root of the repository (that's where currently `inference.py`) should not contain any python files ideally.

**Comment by chani0343** [comment]
Hi 

All feedback addressed:

✅ Test images: Kept only tomato.png and cnw_annotations.jpg
✅ No env/ changes: Restored to match main branch
✅ Project structure: Moved Python files to src/, separated data/ and models/

Ready for merge - all checks passed, no conflicts.

קצר ולעניין!

**Review by lyuzinmaxim — APPROVED** [review]
